File: client-flutter/lib/models/appointment.dart
class Appointment {
  final String id;
  final DateTime startDate;
  final int duration;
  final String text;
  final String trainingTypeId;
  final String trainingTypeName;
  final String locationId;
  final String locationName;
  final String trainerId;
  final String trainerName;
  final int creditsCharged;
  final String status;

  Appointment({
    required this.id,
    required this.startDate,
    required this.duration,
    required this.text,
    required this.trainingTypeId,
    required this.trainingTypeName,
    required this.locationId,
    required this.locationName,
    required this.trainerId,
    required this.trainerName,
    required this.creditsCharged,
    required this.status,
  });

  factory Appointment.fromJson(Map<String, dynamic> json) {
    // Combine date + starttime from NestJS backend
    DateTime parseDate() {
      final date = json['date']?.toString() ?? '';
      final time = json['starttime'] ?? json['time_from'] ?? '00:00';
      if (date.isEmpty) return DateTime(1970);
      try {
        return DateTime.parse('${date}T$time');
      } catch (_) {
        return DateTime.tryParse(date) ?? DateTime(1970);
      }
    }

    final trainer = json['trainer'];
    String trainerName = '';
    if (trainer is Map<String, dynamic>) {
      trainerName =
          '${trainer['firstname'] ?? ''} ${trainer['lastname'] ?? ''}'.trim();
    }

    return Appointment(
      id: json['id']?.toString() ?? '',
      startDate: parseDate(),
      duration: int.tryParse(json['duration']?.toString() ?? '60') ?? 60,
      text: json['notes']?.toString() ?? json['text']?.toString() ?? '',
      trainingTypeId: json['training_type_id']?.toString() ?? '',
      trainingTypeName: json['training_type_name']?.toString() ?? '',
      locationId: json['location_id']?.toString() ?? '',
      locationName: json['location_name']?.toString() ?? '',
      trainerId: json['trainer_id']?.toString() ?? '',
      trainerName: trainerName,
      creditsCharged:
          int.tryParse(json['credits_charged']?.toString() ?? '0') ?? 0,
      status: json['status']?.toString() ?? '',
    );
  }
}

/// Data returned by /api/client/start/:id
class StartData {
  final String firstName;
  final String lastName;
  final int totalCredits;
  final List<Appointment> appointments;
  final String? lastTrainingDate;
  final int trainingsThisWeek;

  StartData({
    required this.firstName,
    required this.lastName,
    this.totalCredits = 0,
    required this.appointments,
    this.lastTrainingDate,
    this.trainingsThisWeek = 0,
  });

  factory StartData.fromJson(Map<String, dynamic> json) {
    final rawAppts = json['appointments'];
    final appointments = <Appointment>[];
    if (rawAppts is List) {
      for (final a in rawAppts) {
        if (a is Map<String, dynamic>) {
          appointments.add(Appointment.fromJson(a));
        }
      }
    }
    return StartData(
      firstName: json['firstName']?.toString() ?? json['firstname']?.toString() ?? '',
      lastName: json['lastName']?.toString() ?? json['lastname']?.toString() ?? '',
      totalCredits: int.tryParse(json['credits']?.toString() ?? '0') ?? 0,
      appointments: appointments,
      lastTrainingDate: json['lastTrainingDate']?.toString(),
      trainingsThisWeek: int.tryParse(json['trainingsThisWeek']?.toString() ?? '0') ?? 0,
    );
  }
}

File: client-ios/SihlClient/Models/Appointment.swift
import Foundation

/// Pendant zu `models/appointment.dart`.
struct Appointment: Identifiable {
    let id: String
    let startDate: Date
    let duration: Int
    let text: String
    let trainingTypeId: String
    let trainingTypeName: String
    let locationId: String
    let locationName: String
    let trainerId: String
    let trainerName: String
    let creditsCharged: Int
    let status: String
    let trainingPlanId: Int?
    let trainingPlanName: String?
}

extension Appointment {
    /// Baut ein Appointment aus einem rohen JSON-Dictionary.
    /// Das Backend liefert Datum und Uhrzeit getrennt ("date" + "starttime").
    init?(json: [String: Any]) {
        guard let idVal = json["id"] else { return nil }
        self.id = "\(idVal)"

        let dateStr = (json["date"] as? String) ?? ""
        let timeStr = (json["starttime"] as? String)
                   ?? (json["time_from"] as? String)
                   ?? "00:00"
        if dateStr.isEmpty {
            self.startDate = Date(timeIntervalSince1970: 0)
        } else {
            let combined = "\(dateStr)T\(timeStr)"
            // Try ISO 8601 first; fall back to date-only
            let isoFmt = ISO8601DateFormatter()
            isoFmt.formatOptions = [.withFullDate, .withTime,
                                    .withColonSeparatorInTime,
                                    .withDashSeparatorInDate]
            if let d = isoFmt.date(from: combined) {
                self.startDate = d
            } else if let d = ISO8601DateFormatter().date(from: dateStr) {
                self.startDate = d
            } else {
                self.startDate = Date(timeIntervalSince1970: 0)
            }
        }

        self.duration        = Int("\(json["duration"] ?? 60)") ?? 60
        self.text            = (json["notes"] as? String) ?? (json["text"] as? String) ?? ""
        self.trainingTypeId  = "\(json["training_type_id"] ?? "")"
        self.trainingTypeName = "\(json["training_type_name"] ?? "")"
        self.locationId      = "\(json["location_id"] ?? "")"
        self.locationName    = "\(json["location_name"] ?? "")"
        self.trainerId       = "\(json["trainer_id"] ?? "")"

        if let trainer = json["trainer"] as? [String: Any] {
            let first = (trainer["firstname"] as? String) ?? ""
            let last  = (trainer["lastname"]  as? String) ?? ""
            self.trainerName = "\(first) \(last)".trimmingCharacters(in: .whitespaces)
        } else {
            self.trainerName = ""
        }

        self.creditsCharged = Int("\(json["credits_charged"] ?? 0)") ?? 0
        self.status         = (json["status"] as? String) ?? ""

        let rawPlanId = json["training_plan_id"] ?? json["trainingPlanId"]
        self.trainingPlanId = Int("\(rawPlanId ?? "")")

        if let name = json["training_plan_name"] as? String {
            self.trainingPlanName = name
        } else if let plan = json["trainingPlan"] as? [String: Any] {
            self.trainingPlanName = plan["name"] as? String
        } else {
            self.trainingPlanName = nil
        }
    }
}

// MARK: - StartData

/// Antwort von `GET /api/client/start/:id` — Pendant zu `StartData` in Dart.
struct StartData {
    let firstName: String
    let lastName: String
    let totalCredits: Int
    let appointments: [Appointment]

    init(json: [String: Any]) {
        self.firstName    = (json["firstName"] as? String) ?? (json["firstname"] as? String) ?? ""
        self.lastName     = (json["lastName"]  as? String) ?? (json["lastname"]  as? String) ?? ""
        self.totalCredits = Int("\(json["credits"] ?? 0)") ?? 0

        if let list = json["appointments"] as? [[String: Any]] {
            self.appointments = list.compactMap { Appointment(json: $0) }
        } else {
            self.appointments = []
        }
    }
}
